Arm64 rebase (work in progress)

swine opened the pull request:
a rebase of the earlier arm64 support from surajjs95/kpatch, tested with proposed kernel changes from https://lore.kernel.org/linux-arm-kernel/20220524001637.1707472-1-madvenka@linux.microsoft.com

More of a heads-up than a pull-request, as I haven't re-tested against (patched) upstream kernel, or baseline clang & gcc yet, just the in-house 5.15ish frankenkernel & custom toolchain, producing well-behaved x86 & arm64 patch modules.
I'll update when tested further against generic kernel/toolchain

While surajjs95's original version worked well with 5.10 kernel codebase (with v13 of madvenka's klp/arm64 changes), and earlier versions of clang, updates to kernel and clang have required changes to elf section namespace & clang local symbol namespace.

The most interesting change since Suraj's version is support for the multiple __patchable_function_entries sections that modern clang emits for arm64 (and riscv too, I hear)

joe-lawrence commented:
Thanks @swine!

BTW, I wonder if development of kpatch-build for arm64 would be easier if we were to merge work to a temporary development branch?  That would facilitate contributions from multiple folks without having to carry the entire history over and over again... and then we just periodically rebase to keep up with master until the work is ready.  Just a thought.

On the PR:
- check out the build tests that are failing (looks pretty minor at first glance)
- how many of the new clang issues would affect x86?  could we spin those out to merge sooner than later?
- do you happen to know what post v5.10 patches Madhavan's latest patchset now depends on?

swine commented:
I have yet to try this against ubuntu/x86 to understand what broke in the build/unit-tests

Probably a better way of layering arm64 support, since @t-msn has published a similar gcc-focused arm64 rebase, is to layer these partially clang-related changes atop t-msn's, if they address similar issues in compatible ways.
I've not reviewed t-msn's changes deeply, while much of it looked familiar from Suraj's earlier work, there are some apparently gcc-namespace (or post-5.15 kernel namespace) related issues that should be merged earlier than my clang-namespace changes, as most of the community is still gcc-centric, and the unit tests are by default (or always?) using gcc

Next layer to merge would be clang/x86 issues, and finally the clang/arm64 issues, including the multiple-__patchable_function_entries ELF sections that arm64/riscv now generate

Re: "what post v5.10 patches Madhavan's latest kernel patchset now depends on" ...

This kpatch patchset has been developed against Madhavan's kernel work, as it evolved from v5.12 thru 5.19, and targeting our 5.10 & 5.15 clang-built frankenkernels.

With 5.15/clang, Madhavan's latest objtool-using 5.19-based kernel changes backported easily with 26 supporting objtool backports, and a few other kernel cherry-picks, bridging the 5.15..5.19 gap. This is working reliably on 5.15, passing all our kpatch/klp tests

We're currently backporting this to 5.10, where Madhavan's v13 kernel patches were working nicely (the pre-objtool version relying on SYM_CODE/SYM_FUNC markings), with a few supporting backports in ftrace/etc infrastructure.
But backporting Madhavan's latest objtool-based patchset to 5.10 is still a work in progress - it still needs the ftrace/etc support, but also needs about 70! backports of the objtool evolution between 5.10..5.15.
Most of these are clean cherry-picks into a pure 5.10 tree, but we're still debugging - builds for both x86/arm64, but patches fail to apply, so something is broken, probably the objtool merges.

I'm sure the arm64 objtool changes could be backported without most of the intervening 5.10...5.19 evolution of largely-x86-specific functionality, but since we aggressively backport objtool changes for retpoline & ebpf support anyway, it seems cleaner to bring the whole baseline up closer to upstream, than to diverge further, forking objtool for 3 distinct needs.

The objtool-using patchset is certainly the preferred kernel solution - verifying sane frame-pointer semantics rather than relying on developers to keep SYM_CODE/SYM_FUNC markings accurate.

t-msn commented:
Hello, @joe-lawrence 
To clarify, my rebase branch @swine mentioned is [here](https://github.com/t-msn/kpatch/commits/arm64_rebase_202209), which I described in the original PR [comment ](https://github.com/dynup/kpatch/pull/1236#issuecomment-1247637140))
My rebase is another rebase to master branch based on your rebase and does not care clang build at this point.

> swine's comment
> I've not reviewed t-msn's changes deeply, while much of it looked familiar from Suraj's earlier work,

Basically I tried to keep the original patches' structure as possible.

I checked that the differences of rebase at the commit "kpatch-build: Enable ARM64 support" (i.e. [mine](https://github.com/t-msn/kpatch/commit/ce0d0577ae6eb71a0073dfa74c23b2bbe369303b) and [swine's](https://github.com/dynup/kpatch/pull/1302/commits/9f5c5ff28863bda45c60d3417293d06e01b0c0cb)).
I think one notable difference is that swine's version addresses .static_call_sites sections which didn't care  in the original patch.
After that commit, my rebase branch contains joe-lawrence's unit tests commit for arm64 and some small fixes.

> joe-lawrence's comment
> BTW, I wonder if development of kpatch-build for arm64 would be easier if we were to merge work to a temporary development branch? That would facilitate contributions from multiple folks without having to carry the entire history over and over again... and then we just periodically rebase to keep up with master until the work is ready. Just a thought.

I agree this so that people can easily test and develop using the shared code (i.e. rebased original PR + unit tests). I'm not sure what is the best way but for me it is both ok to create another PR with my rebase code or to split this PR up to "Enable ARM64 support".  Any thoughts?

joe-lawrence commented:
> I have yet to try this against ubuntu/x86 to understand what broke in the build/unit-tests
> 
> Probably a better way of layering arm64 support, since @t-msn has published a similar gcc-focused arm64 rebase, is to layer these partially clang-related changes atop t-msn's, if they address similar issues in compatible ways. I've not reviewed t-msn's changes deeply, while much of it looked familiar from Suraj's earlier work, there are some apparently gcc-namespace (or post-5.15 kernel namespace) related issues that should be merged earlier than my clang-namespace changes, as most of the community is still gcc-centric, and the unit tests are by default (or always?) using gcc

Yes, unit tests are cached gcc-generated original/patched object files.  We can add clang generated object files if there are issues specific to its toolchain.  Curation of the unit test object files is unfortunately a manual process, but at least that makes it easy to add "clang-test-foo.o" files to the suite.

Regarding the two MRs, rebases of rebases, etc. are starting to get hard to track... so what about this strategy largely based on your suggestion:
1. @swine / @joe-lawrence opens a new PR with clang/x86 related issues from this set.  We can review and merge these independently and sooner than any of the arm64 related ones.  BTW, we have hit some of these issues internally (like .llvm* sections, .shstrtab, etc.) but haven't had time to fix them, hence my interest.
2. @joe-lawrence creates a new aarch64-development branch based on @t-msn's rebase.  As noted, this branch looks cleanest with respect to handling arch-specific C code from the onset.
2a. We can periodically rebase and force push to keep up with master
2b. Fixes to these commits can be squashed as part of periodic rebase (2a).
2c. New development can just add new commits as usual
3. @swine opens a new PR on top of (2) with toolchain-generic arm fixes like: ("harvest paravirt/alt_instr sizes for 5.15"), ("arm64 leaf-function fix"), ("kpatch-cc skip arch/arm64/kernel/vdso*/*")  These would follow step (2c) above so that gcc based development can benefit, too.
4. @swine opens a new PR on top of (2) with clang/specific arm64 fixes.

> 
> Re: "what post v5.10 patches Madhavan's latest kernel patchset now depends on" ...
> 
> This kpatch patchset has been developed against Madhavan's kernel work, as it evolved from v5.12 thru 5.19, and targeting our 5.10 & 5.15 clang-built frankenkernels.
> 
> With 5.15/clang, Madhavan's latest objtool-using 5.19-based kernel changes backported easily with 26 supporting objtool backports, and a few other kernel cherry-picks, bridging the 5.15..5.19 gap. This is working reliably on 5.15, passing all our kpatch/klp tests
> 
> We're currently backporting this to 5.10, where Madhavan's v13 kernel patches were working nicely (the pre-objtool version relying on SYM_CODE/SYM_FUNC markings), with a few supporting backports in ftrace/etc infrastructure. But backporting Madhavan's latest objtool-based patchset to 5.10 is still a work in progress - it still needs the ftrace/etc support, but also needs about 70! backports of the objtool evolution between 5.10..5.15. Most of these are clean cherry-picks into a pure 5.10 tree, but we're still debugging - builds for both x86/arm64, but patches fail to apply, so something is broken, probably the objtool merges.
> 
> I'm sure the arm64 objtool changes could be backported without most of the intervening 5.10...5.19 evolution of largely-x86-specific functionality, but since we aggressively backport objtool changes for retpoline & ebpf support anyway, it seems cleaner to bring the whole baseline up closer to upstream, than to diverge further, forking objtool for 3 distinct needs.
> 
> The objtool-using patchset is certainly the preferred kernel solution - verifying sane frame-pointer semantics rather than relying on developers to keep SYM_CODE/SYM_FUNC markings accurate.

Thanks for the details!  I haven't looked at Madhavan's latest few versions, so I am asking as rhel-9 franken-kernel backports are starting to get interesting, too.

t-msn commented:
> Regarding the two MRs, rebases of rebases, etc. are starting to get hard to track... so what about this strategy largely based on your suggestion:

I have no objections for this approach. Thank you very much for you effort.

> 2. @joe-lawrence creates a new aarch64-development branch based on @t-msn's rebase. As noted, this branch looks cleanest with respect to handling arch-specific C code from the onset.

It might be easier for you to rebase by yourself than using my rebase

swine commented:
See updated description - now based on @t-msn changes laid over master, may be candidate for a temporarily arm64-specific branch until it's shown harmless to other platforms

t-msn commented on kpatch-build/create-diff-object.c:
```diff
@@ -3659,29 +3660,51 @@ static void kpatch_alloc_mcount_sections(struct kpatch_elf *kelf, struct kpatch_
+			if (multi_pfe)
+				sec->sh.sh_link = 0; /* set later */
+			else
+				sec->sh.sh_link = text_idx;
```
I saw the following error with gcc12 for integration test (data-new etc.):

> ld: __patchable_function_entries has both ordered and unordered sections

Bisect points this commit. With additional log messages, I found text_idx is 0 and it causes the problem.
Should we set sh_link later like when multi_pfe is true or set 1 unconditionally just like before this commit?
(I'm looking this comment...: https://github.com/dynup/kpatch/pull/1236#issuecomment-1040197012)

swine replied:
I think the issue was that multi_pfe was getting set too late.
Next push should resolve that

ryncsn replied:
Hi @swine, I'm still seeing
```
ld: __patchable_function_entries has both ordered and unordered sections
```
with gcc12 and your latest push (2023.1.16, d01cbc2), set sh_link to 1 unconditionally fixed the issue for me.

joe-lawrence reviewed (commented):
Thanks for updating, @swine.  See a few inline comments on various bits of code.

Then to save dragging these through future rebases, can you spin out the following patches as their own PR?  They don't appear to be arm64-arch dependencies:

- ("kpatch-build: determine ARCH from $VMLINUX ")
- ("allow env override of toolchain")
- ("Read backported klp capabilities from <linux/livepatch.h>")
- ("simplify CC/LD patterns")
- ("add --make-opt, --cdo-opt")

swine commented on kpatch-build/kpatch-build:
```diff
@@ -642,6 +641,20 @@ fi
+(*aarch64*) ARCH=arm64;;
+(*) die "Kernel ${VMLINUX:-$BASE/vmlinux} has Unsupported Architecture (${ARCH})";;
+esac
+
```
exactly - I'll split off as a separate PR, to be discussed and resolved as an orthogonal issue.
It was bundled in here because @t-msn & I reworked @surajjs95 changes from '#ifdef __arch64__' to 'switch(kelf->arch)', and this extended the cross-arch transformation

swine replied:
but when this comes back as cross-architecture support proposal, it will use elf/uname arch names like aarch64, not try to subvert kpatch-build tradition by changing the arch names to kernel makefile versions like arm64+x86_64

swine commented on kpatch-build/kpatch-build:
```diff
@@ -357,6 +357,9 @@ find_special_section_data() {
+		"arm64")
+			check[a]=true					# alt_instr
+			;;
 	esac
```
yes, there may be more dup code from rebasing my rebase over @t-msn's, I'll explore

swine commented on kpatch-build/kpatch-build:
```diff
@@ -57,12 +57,14 @@ APPLIED_PATCHES=0
+: "${READELF:="${CROSS_COMPILE:-}readelf"}"
+: "${OBJCOPY:="${CROSS_COMPILE:-}objcopy"}"
+: "${MAKE:=make}"
 
```
Yes, this is one of the changes for working with cross-compile, and using kernel-build-env designated toolchain, rather than the defaults a desktop build env might provide.

It's a layered default scheme, where CROSS_COMPILE bulk-overrides the tools that the default names might resolve from $PATH, but per-tool overrides can special-case that generic override.
Particularly useful (for example) where LLVM toolchain employed, but llvm's readelf didn't behave as required, and needed a wrapper script. Haven't documented that as I'm still unsure whether that's a local toolchain quirk, or a generic issue.

-: "${LLD:="${CROSS_COMPILE:-}ld.lld"}"
+# use CROSS_COMPILE as kernel/Makefile does, or the per-tool overrides if specified
+: set default to "${LLD:="${CROSS_COMPILE:-}ld.lld"}"
...

Yes, this also needs documenting in doc/INSTALL.md - none of the per-distro sections need mention these generalizations, for each distro will build its kernel with the tools its dev-environment presents in $PATH. But for cross-compiling, whether for a different arch in the same distro family, or for a wholly different scenario, these are useful.

swine commented on kpatch-build/kpatch-cc:
```diff
@@ -23,7 +23,7 @@ if [[ "$TOOLCHAINCMD" =~ ^(.*-)?gcc$ || "$TOOLCHAINCMD" =~ ^(.*-)?clang$ ]] ; th
 			[[ "$obj" = */.tmp_mc_*.o ]] && break;
 
 			[[ "$obj" = */.tmp_*.o ]] && obj="${obj/.tmp_/}"
-			relobj=${obj##$KPATCH_GCC_SRCDIR/}
```
These were for calming shellcheck, possibly a different version of shellcheck than github uses.
Having been nagged by github's shellcheck, I updated it locally and resolved all its quibbles

swine commented on kpatch-build/create-diff-object.c:
```diff
@@ -579,13 +579,18 @@ static void kpatch_compare_correlated_nonrela_section(struct section *sec)
+#ifndef SHF_INFO_LINK
+# define SHF_INFO_LINK 0x40UL
+#endif
+
```
I'll update comments - I'd run into an extremely ancient version in one of our build environments when adding this workaround, even before Suraj's changes landed. If I can't find evidence it's still needed locally, I'll drop it completely. Anyway it doesn't belong in the arm64 branch

swine replied:
Dropping this - I think it was addressing a consequence of using earlier create-diff-object before Suraj's fix to the sh.sh_link setup, so incoming objects were corrupted

swine commented on kpatch-build/create-diff-object.c:
```diff
@@ -3998,27 +4051,34 @@ static void kpatch_find_func_profiling_calls(struct kpatch_elf *kelf)
+			list_for_each_entry(sec, &kelf->sections, list) {
+				if (strcmp(sec->name, "__patchable_function_entries"))
+					continue;
+				if (multi_pfe && sym->sec->pfe != sec)
```
good catch! - multi_pfe scan should be ahead of it's use.
This may explain some other problems.

swine commented on kpatch-build/create-diff-object.c:
```diff
@@ -70,6 +70,7 @@ enum subsection {
 enum loglevel loglevel = NORMAL;
 
 bool KLP_ARCH;
+bool multi_pfe;
```
It's possible that a given toolchain's _intent_ to provide multiple __pfe is not apparent through examining a single ELF, if that ELF has only one patchable function.
So both must be examined before the requirement to emit multi-__pfe style can be determined.
Hence it's not a property of a kelf object, but it could be a property handled within kpatch-elf.c, though with no _use_ made of it there, it seems uglier than the current state.

When @surajjs95 first pushed changes, both his gcc, and google's clang, emitted just a single __pfe, and the code worked for 5.10 in both environments.
Now, latest version of both emit multiples.

From following the discussion of this departure from the old __mcount_loc, thru single-pfe, to the current multi-pfe, in both gcc & clang, I haven't seen a clear signal from the toolchain on which approach is supported, apart from just counting the pfe sections. So that seems to be the best way of accommodating the difference.

Futures?
Whether __pfe or __mcount_loc is used seems to be not so much a per-arch decision as a toolchain decision, given the rationale given by maskray when championing this -- single section demands that whole section be pulled in during link/load when only partial contents needed -- consider pfe entries for .weak functions, where a generic set of fallback C-coded implementations exists to fill the gaps in a per-arch set of optional optimized functions.

With that trajectory in mind, perhaps this might one day become
  enum pfe_style { mcount_loc=0, single_pfe=1,multi_pfe=2};
But later.

swine replied:
anyway, I've moved the setting of multi_pfe before its first use, which eliminates the real cause of the problem which manifested as if something was not setting/defining SHF_LINK

jpoimboe commented on kpatch-build/create-diff-object.c:
```diff
@@ -788,6 +838,84 @@ static bool kpatch_line_macro_change_only(struct kpatch_elf *kelf,
 	return true;
 }
 
+static bool _kpatch_line_macro_change_only_aarch64(struct kpatch_elf *kelf,
```
Do we really need an arm64-specific version?  At least from a quick glance I don't see why the generic version wouldn't work.

jpoimboe commented on kpatch-build/create-diff-object.c:
```diff
@@ -1042,9 +1170,14 @@ static void kpatch_correlate_symbols(struct list_head *symlist_orig,
+			     !strncmp(sym_orig->name, ".Ltmp", 5)))
+				continue;
+
+			if (kpatch_is_mapping_symbol(kelf_orig, sym_orig))
```
coding style

jpoimboe commented on kpatch-build/create-diff-object.c:
```diff
@@ -3534,34 +3673,118 @@ static void kpatch_create_callbacks_objname_rela(struct kpatch_elf *kelf, char *
+		for (nr = 0; nr < copies; nr++) {
+			sec = create_section_pair(kelfout,
+				"__patchable_function_entries",
+				sizeof(void*), entries);
```
Here and elsewhere, please align the function arguments

jpoimboe commented on kpatch-build/create-diff-object.c:
```diff
@@ -3640,6 +3890,20 @@ static void kpatch_create_mcount_sections(struct kpatch_elf *kelf)
+					       "__patchable_function_entries");
+			if (!sec)
+				ERROR("cannot retrieve pre-allocated __pfe #%d\n",
+				       nr);
```
All on a single line please

jpoimboe reviewed (commented):
In addition to the individual code comments:

- Please no version changelog information in the commit logs
- Commit 0b09a5755291 ("create-diff-object: Create __patchable_function_entries section for aarch64") updates `test/unit/objs`.  I don't think that was intentional?

swine commented on kpatch-build/create-diff-object.c:
```diff
@@ -3802,12 +4072,39 @@ static void kpatch_find_func_profiling_calls(struct kpatch_elf *kelf)
+
+			list_for_each_entry(sec, &kelf->sections, list) {
+				if (strcmp(sec->name, "__patchable_function_entries"))
+					continue;
```
it's a toolchain evolution, possibly coming in gcc too, where the section-per-function allows garbage collection of these __pfe sections, when the functions they annotate are not used. Consider linking with a lib, most of the funcs of which are never referenced. Without this all the __pfe entries would be survive, even if the funcs they refer to do not.

See https://www.mail-archive.com/gcc@gcc.gnu.org/msg90391.html for context.

I'm working with google-internal clang which probably accumulates features by backport in a different order to upstream, so even if it was clang-specific, it's better to diagnose this behavior by counting the incoming __pfe sections than switching on some version comparison. My hope is that code works unmodified when it encounters a gcc which implements this behavior, but I haven't yet found one

swine replied:
See "On the __patchable_function_entries issues..." below.

It's actually an _anti-bloat_ issue, enabling $(LD) to garbage-collect using only the sh_link dependencies and property flags, without embedding logic that ties meaning to symbol or section names. With Clang-LTO (or gcc's analogue) the final link has all its balls in the air at once, and any means of culling the flock is helpful. By the time the final ELF lands, all these zillions of sections have coalesced under vmlinux.lds.S into the usual suspects

swine commented:
in the final kernel-or-object link all the surviving __pfe sections are
collapsed into one.

I'll explore further with maskray on which versions this lands in, and
whether I've got the whole issue wrong

-pete


On Wed, Nov 2, 2022 at 8:50 AM Josh Poimboeuf ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In kpatch-build/create-diff-object.c
> <https://github.com/dynup/kpatch/pull/1302#discussion_r1011975531>:
>
> >  			continue;
>
>  		switch(kelf->arch) {
> +		case AARCH64: {
> +			struct section *sec;
> +
> +			list_for_each_entry(sec, &kelf->sections, list) {
> +				if (strcmp(sec->name, "__patchable_function_entries"))
> +					continue;
>
> From a kernel standpoint this seems more like an anti-feature, since IIUC
> it adds hundreds of thousands of extra sections and bloats the object file
> size, with no benefit for the kernel. It would be nice if there were a flag
> to toggle this behavior.
>
> Also I'm still not able to recreate it with upstream Clang, maybe I'm
> doing something wrong?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/dynup/kpatch/pull/1302#discussion_r1011975531>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAH4XE6ISZR6YF3VQEPRR4TWGKEUPANCNFSM6AAAAAAQNZK4AY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

swine commented:
Rebased again, addressed most? of Josh's comments

(apologies for long delay - I thought I'd pushed a rebase/cleanup weeks ago - I was preening comments when wife broke her leg, forgot I'd left it unsent)

One unit-test failure - ppc64le/1102, which I'd assume is because my submodule is out of date, haven't yet tried to freshen with recent changes

swine commented:
I'm still hunting the unit-test failures above

On the __patchable_function_entries issues, there's a maskray backgrounder at http://maskray.me/blog/2020-02-01-fpatchable-function-entry (chinese, google-translate renders it well in english).

Basically, garbage collection of unused sections is simplified when these __pfe markings are attached by elf sh_link, so they disappear when the .text they refer to disappears. Combining them would mean they would persist until their last target disappears.
Both gcc & clang are implementing this (stable from clang-13 & gcc-5.1 if I read the logs correctly) as a response to -fpatchable-function-entry=N,M, which injects N-byte NOP at entry, M-byte at exit. This renders the external recordmcount pass obsolete.

Per-arch makefiles typically add CFLAGS += -DCC_USING_PATCHABLE_FUNCTION_ENTRY when using this, as done for (arm64 parisc riscv loongarch)

Just to confuse things, 6.0's commit 9440155ccb ("ftrace: Add HAVE_DYNAMIC_FTRACE_NO_PATCHABLE") says "x86 will shortly start using -fpatchable-function-entry for purposes other than ftrace", so if we wanted to remove the arm64-specific logic here, it would be better to check for the absence of __mcount_loc, rather than the presence of __pfe.
But for now explicit switch on elf->arch is cleanest - as other archs are made kpatchable, they can extend that.

The "zillions of __pfe sections" issue is not a general kernel problem - there's one for each .textish section, and we're only getting zillions because -ffunction-sections is creating zillions of .text.*, each of which _may_ get a __pfe section. They're all collapsed together at link time, unless their .textish parent has disappeared in link-time optimization.

For patch-creation, there's some duplication in that any __pfe and .rela__pfe sections output may generate individual .symtab entries, even though their section headers could share these constant symtab entries, but that's also seen in compiler-generated *.o. That should also collapse during .ko creation, by
```scripts/module.lds.S:	__patchable_function_entries : { *(__patchable_function_entries) }``` or similar in vmlinux.lds.S

swine commented:
unit test failures are resolved by merging the latest ppc64le changes to test/unit/objs

I'm still pondering whether Josh's suggestion of merging _kpatch_line_macro_change_only_aarch64() with the other-arch case makes sense - probably does, as the aarch64 skipping of most of the logic is not an _architecture_ decision, but an implementation decision - other archs may well adopt similar solutions to the removal of __LINE__ inline constants from WARN/BUG/etc, so it's likely to simplify future ports which also avoid the issues

t-msn commented:
Thank you for updating patches.

I tried clang build with the [latest kernel update based on 6.1](https://lore.kernel.org/linux-arm-kernel/20230202074036.507249-1-madvenka@linux.microsoft.com/) but got error with clang 15.0.7.
I think you are using custom llvm/clang. So do you know if (or when) this series works with upstream llvm/clang?

For gcc, with the fix for the problem in https://github.com/dynup/kpatch/pull/1302#discussion_r995662006, all integration tests in linux-5.18.0 directory pass fine (except module test, which needs another [patch in kernel](https://lore.kernel.org/linux-arm-kernel/20211103210709.31790-1-surajjs@amazon.com/)) for above linux series with gcc 12.2.1.
(btw, to kpatch 6.1 kernel I need to cherry-pick https://github.com/dynup/kpatch/commit/934b3b3c0c12be2435f8edbe484c2696dda93864)

swine commented:
> I tried clang build with the [latest kernel update based on 6.1](https://lore.kernel.org/linux-arm-kernel/20230202074036.507249-1-madvenka@linux.microsoft.com/) but got error with clang 15.0.7. 

Yes, I had until now been using our internal custom kernel & tools, and quick tests with kernel.org's v6.1

I'm now cross-testing with upstream 6.1 & a more mainstream toolset (ubuntu's clang 15.0.6), and I'm seeing related issues.
I'll try to widen testing before next update.
Looks like github actions are restricted to ubuntu-latest, but I'd like to validate on RHEL and other platforms, with both gcc & clang

ZzzMao commented:
I have some different understanding for the `multi_pfe` and the separte of the `kpatch_create_mcount_sections` part. 

Setting the `sh_link` to `0` if definitely wrong. And set it to `1` is also not correct. It is a little like the UB in C: "yes, it works now. But it might break in the future". The sh_link of a pfe section should always link back into one of the text sections.

For the `multi_pfe`, in ARM64, there should always be multiple __patchable_function_entries sections when there are multiple functions in the .c file. 
It is a bug in GCC 11/12, where the compiler will only generate a single pfe section. The bug has been fixed in GCC 13.1.
(@MaskRay helped me confirmed this bug here: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=110729)
And in clang, there is no such bug.

@jpoimboe, I want to try to answer your previous question on the pfe section's sh_flag and sh_link.
My understanding is: yes, we can create the pfe sections in the `create_mcount` function without splitting it into two functions. And we can do it correct. 
Let's ignore the GCC bug first. Every function should have its own pfe section. Though these pfe sections would have the same name. Each pfe section should set its `sh_link` to the corelated .text section and with `SHF_LINK_ORDER` enabled. 
As the single pfe behavior is a bug in older version of GCC, I think we do not need **simulate** this bug in `kpatch-build`. Instead, we can just do the correct thing.
MaskRay has another blog about the `SHF_LINK_ORDER` here (and in English this time): https://maskray.me/blog/2021-01-31-metadata-sections-comdat-and-shf-link-order


Besides, there is a bug in the current code. The pfe is not initialized before the `populate_mcount` function. But it is used later in `kpatch_find_func_profiling_calls` when the `multi_pfe` is set to true.

I have a commit to fix all these issues here: 
https://github.com/ZzzMao/kpatch/commit/d454fdb5ea253efa88e62352253393bce7edbf11
I have also performed some end to end test based on Madhavan's [orc_v3](https://github.com/madvenka786/linux/tree/orc_v3) change on kernel v6.1 with `gcc-12` and `clang-14`. Both works fine with my code.

@swine, if you are fine with that commit, can I submit a PR to merge it to your `swine:arm64`?

ZzzMao commented on kpatch-build/create-diff-object.c:
```diff
@@ -3715,6 +3830,42 @@ static void kpatch_create_mcount_sections(struct kpatch_elf *kelf)
+			sec->sh.sh_flags |= SHF_WRITE | SHF_LINK_ORDER;
+			/* We will reset this sh_link in the reindex function. */
+			sec->sh.sh_link = 0;
+
```
A FYI for this change: 
Thought I am emitting multiple "pfe" sections for every single functions here. In the later step in `kpatch-build`, the `$(LD) -r output.o <input.o files>` might still merge these pfe sections into one. At least this is current (v2.41) behavior for `ld` and `ld.gold` . 
For `ld.lld`, it will keep these multiple pfe sections as separated. 

But I think it is not a big issue here. The pfe sections will be squashed into one anyway in the finally ko file when building the `/kmod/patch`.

ZzzMao commented on kpatch-build/create-diff-object.c:
```diff
@@ -228,6 +230,24 @@ static struct rela *toc_rela(const struct rela *rela)
+ * as they often all have the same name either "$x" to mark the start of code
+ * or "$d" to mark the start of data.
+ */
+static bool kpatch_is_mapping_symbol(struct kpatch_elf *kelf, struct symbol *sym)
```
I think we can change the function signature to `static bool kpatch_is_mapping_symbol(enum architecture arch, struct symbol *sym)`. 
We only need the arch information, not the whole elf here.

There are only 2 usage for this function, `kpatch_replace_sections_syms` and `kpatch_correlate_symbols`.
There would be no difference for the `kpatch_replace_sections_syms`, as it is already a very complex function.

But for the `kpatch_correlate_symbols`, I feel it should have a similar signature as the `kpatch_correlate_sections`. But in the current code, it takes two `kpatch_elf*` as input parameters, only for getting this arch information. 

I know we have to have some compromise due to problem complexity. But I think, comparing to have two `kpatch_elf*`,  `kpatch_correlate_symbols(struct symbol *sym_orig, struct symbol *sym_patched, enum architecture arch)` looks more reasonable to me.

@swine, If you are fine with this change, I can create another PR to your branch later.

swine commented:
Yes, that's looking much cleaner

earlier code was done when gcc + clang had different behaviors, but now
that the direction of evolution is clear (both will emit multiple __pfe),
this is less experimental, more generalizable across all archs because we
know that the concept is arch-independent, and can clean up function
signatures, rather than striving for minimal change

-pete


On Fri, Sep 1, 2023 at 3:15 PM Ziwei Mao ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In kpatch-build/create-diff-object.c
> <https://github.com/dynup/kpatch/pull/1302#discussion_r1313583807>:
>
> > @@ -228,6 +230,24 @@ static struct rela *toc_rela(const struct rela *rela)
>  				   (unsigned int)rela->addend);
>  }
>
> +/*
> + * Mapping symbols are used to mark and label the transitions between code and
> + * data in elf files. They begin with a "$" dollar symbol. Don't correlate them
> + * as they often all have the same name either "$x" to mark the start of code
> + * or "$d" to mark the start of data.
> + */
> +static bool kpatch_is_mapping_symbol(struct kpatch_elf *kelf, struct symbol *sym)
>
> I think we can change the function signature to static bool
> kpatch_is_mapping_symbol(enum architecture arch, struct symbol *sym).
> We only need the arch information, not the whole elf here.
>
> There are only 2 usage for this function, kpatch_replace_sections_syms
> and kpatch_correlate_symbols.
> There would be no difference for the kpatch_replace_sections_syms, as it
> is already a very complex function.
>
> But for the kpatch_correlate_symbols, I feel it should have a similar
> signature as the kpatch_correlate_sections. But in the current code, it
> takes two kpatch_elf* as input parameters, only for getting this arch
> information.
>
> I know we have to have some compromise due to problem complexity. But I
> think, comparing to have two kpatch_elf*, kpatch_correlate_symbols(struct
> symbol *sym_orig, struct symbol *sym_patched, enum architecture arch)
> looks more reasonable to me.
>
> @swine <https://github.com/swine>, If you are fine with this change, I
> can create another PR to your branch later.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/dynup/kpatch/pull/1302#pullrequestreview-1607715363>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAH4XE6HNY2A52U7YLPGD73XYJNBZANCNFSM6AAAAAAQNZK4AY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

heuza commented on kpatch-build/create-diff-object.c:
```diff
@@ -3754,114 +3753,68 @@ static void kpatch_create_callbacks_objname_rela(struct kpatch_elf *kelf, char *
 	struct symbol *sym;
 	struct rela *rela, *mcount_rela;
 	void **funcs;
+	bool pfe_per_function;
```
pfe_per_function should be default false.
Otherwise, it will be used without initialized when arch !=AARCH64

ZzzMao replied:
Thanks for catch this. I was my mistake. But sorry, I do not have permission to push the "from branch". I will try to send another PR to swine's repo later.

MaskRay commented on kpatch-build/kpatch-elf.c:
```diff
@@ -88,17 +88,29 @@ struct section *find_section_by_index(struct list_head *list, unsigned int index
-	list_for_each_entry(sec, list, list)
-		if (!strcmp(sec->name, name))
-			return sec;
+	if (!list || !list->next || !name)
```
(I know nearly nothing about kpatch. Drive-by comment...)

The condition looks strange. I believe it's not this function job to check `!list || !name`. It should be a precondition calling this function.

`!list->next` can be subsumed by the loop below.

iMonZ commented:
Hey @swine any news? Do you need help in anything? What do we miss befor the merge?

joe-lawrence commented:
(assigned to hold off the stale PR bot)

anatasluo commented:
Can anyone continue this work? We are in an urgent need to use kpatch in arm64.

puranjaymohan commented:
Hi,
If there is no one looking at this, I would like to send the next version of this PR with the requested changes. It would be nice to have ARM64 support merged sooner.

ZzzMao commented:
I think the issue for this PR is not "what is not finished in kpatch". 
But as in the "arm64-upstream-prerequisites.md" mentioned, the upstream kernel does not support livepatch APIs for ARM64 yet. For anyone who wants to use this ARM64 version of kpatch, they need to provide a "klp-enabled" kernel image with some private patches.

heuza commented:
IIRC, upstream propose implementing arm64 livepatch based on SFrame instead of using objtool to reverse engineer binary files.

puranjaymohan commented:
> I think the issue for this PR is not "what is not finished in kpatch". But as in the "arm64-upstream-prerequisites.md" mentioned, the upstream kernel does not support livepatch APIs for ARM64 yet. For anyone who wants to use this ARM64 version of kpatch, they need to provide a "klp-enabled" kernel image with some private patches.

I am aware of this, we are carrying these kernel patches downstream and use it with the patches in this PR for kpatch. As kpatch is a users-pace tool, this PR can be merged even without kernel support. The kernel work can go on separately.

joe-lawrence commented:
I had proposed a while back that we at least merge this (or perhaps a rebased MR) to a development branch.  Would that be the best way forward until the kernel-side support is implemented for arm64?

puranjaymohan commented:
@joe-lawrence 
> I had proposed a while back that we at least merge this (or perhaps a rebased MR) to a development branch. Would that be the best way forward until the kernel-side support is implemented for arm64?

What is the drawback of merging this support without the kernel-side support is merged? The kernel side is only lacking a way to do reliable stack walking on arm64 which should be implemented using sframe (planned). I feel that Kpatch can merge this without the kernel support as there is no abi between arm64 livepatch in the kernel and arm64 kpatch.

Mainline or a development branch whatever is decided, I am happy to re-spin this PR with a rebase and also inculcating the comments on this PR.

mihails-strasuns commented:
+1, AFAICS merging this before the kernel support has no serious drawbacks and 2 important benefits:

1) arm64 unit tests will start running on master branch, ensuring there is no bitrot over time
2) anyone currently using custom kernel will already be able to use kpatch as-is

Worst case scenario I can imagine is that kernel support will require adjustments for this implementation but that seems neither likely nor very costly.

joe-lawrence commented:
Alright, if some one wants to open a fresh MR with the latest-and-greatest kpatch-build changes, that would be great.  A few suggestions:
* arm64 support should be marked as under-development in both the README.md and kpatch-build program
* outstanding issues should be summarized at the top of the MR
* compiler requirements clearly spelled out, as well as kernel patches against the latest (or specific) upstream kernel release need to be listed

mihails-strasuns commented:
Hello, a small update on this: we have started working on bringing the series up to date but there are not estimates yet. Need to rebase internal kernel support patches to have something publicly visible and do a fair a bit of testing to make sure updated kpatch build still works for all our cases. Just FYI so there is no impression this is abandoned.

dylanbhatch commented:
Hi @mihails-strasuns, @swine, @t-msn. Is this PR still active, or is development now switching to klp-build?

@heuza and I have been investigating a way to get create-diff-object to work with DYNAMIC_FTRACE_WITH_CALL_OPS enabled and I think this might work as a solution: dylanbhatch/kpatch@a2e500d. Could this possibly be cherry-picked into this PR (forgive my lack of github knowledge)?

puranjaymohan commented:
@dylanbhatch I am planning to rebase this PR on the latest master and create a new PR. I will add your patch to it.

Thanks

puranjaymohan commented:
Hi I have opened a new PR that is based on later master: https://github.com/dynup/kpatch/pull/1439 and is supposed to work with the `sframe` based unwinder

please let me know if I have missed moving a commit from here to there.
